From: Len Brown <len.brown@intel.com>
Subject: cpuidle: add cpuidle_unregister_driver() error check
References: fate#311818,fate#311942
Git-commit: 329dc1a14c6ddfd71b0ecff1e2458f50258106cf


Signed-off-by: Thomas Renninger <trenn@suse.de>

upstream: c0d64cb031c21f163a0ec15cf10844bcf0ceedcf

Assure that cpuidle_unregister_driver() will not clobber
the registered driver if unregistered by somebody else.

Signed-off-by: Len Brown <len.brown@intel.com>

 drivers/cpuidle/driver.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/cpuidle/driver.c b/drivers/cpuidle/driver.c
index 2257004..826b5c0 100644
--- a/drivers/cpuidle/driver.c
+++ b/drivers/cpuidle/driver.c
@@ -45,8 +45,11 @@ EXPORT_SYMBOL_GPL(cpuidle_register_driver);
  */
 void cpuidle_unregister_driver(struct cpuidle_driver *drv)
 {
-	if (!drv)
+	if (drv != cpuidle_curr_driver) {
+		WARN(1, "invalid cpuidle_unregister_driver(%s)\n",
+			drv->name);
 		return;
+	}
 
 	spin_lock(&cpuidle_driver_lock);
 	cpuidle_curr_driver = NULL;
